lib/cfa/augeas_parser/writer.rb

```diff
@@ -0,0 +1,289 @@
+module CFA
+  # Goal of this class is to write back to augeas data stored in {AugeasTree}.
+  # It tries to only required changes, as augeas keeps inside flag if it is
+  # modified and if not modified, then part of file is kept untouched.
+  # @note internal only, unstable API
+  # @private
+  class AugeasWriter
+    # @param aug result of Augeas.create
+    def initialize(aug)
+      @aug = aug
+    end
+
+    # write to augeas data in tree to given prefix
+    # @param prefix [String] where to write tree in augeas
+    # @param tree [CFA::AugeasTree] tree to write
+    def write(prefix, tree, top_level: true)
+      @lazy_operations = LazyOperations.new(aug) if top_level
+      tree.all_data.each do |entry|
+        located_entry = LocatedEntry.new(tree, entry, prefix)
+        process_operation(located_entry)
+      end
+      @lazy_operations.run if top_level
+    end
+
+  private
+
+    # AugeasEntry togethere with information about its location and few
+    # helper methods to detect siblings.
+    class LocatedEntry
+      attr_reader :prefix
+      attr_reader :entry
+      attr_reader :tree
+
+      def initialize(tree, entry, prefix)
+        @tree = tree
+        @entry = entry
+        @prefix = prefix
+      end
+
+      def orig_key
+        entry[:orig_key]
+      end
+
+      def path
+        return @path if @path
+        return nil unless orig_key
+
+        @path = @prefix + "/" + orig_key
+      end
+
+      def key
+        return @key if @key
+
+        @key = @entry[:key]
+        @key = @key[0..-3] if @key.end_with?("[]")
+        @key
+      end
+
+      # @return [LocatedEntry, nil] preceding entry that already exist in augeas
+      # tree or nil if it do not exists.
+      def preceding_existing
+        preceding_entry = preceding_tree.reverse.find do |entry|
+          entry[:operation] != :add
+        end
+
+        return nil unless preceding_entry
+
+        LocatedEntry.new(tree, preceding_entry, prefix)
+      end
+
+      # @return [true, false] returns true if there is in augeas tree any
+      # existing entry
+      def any_following?
+        following_tree.any? { |e| e[:operation] != :remove }
+      end
+
+      # @return[AugeasTree] returns augeas tree nested under entry. If there
+      # is not such tree, it create empty one.
```

Comment:
well, this is significant difference. For value, nil is valid and it means no value. On other hand for tree, empty tree means no value. So I use it as NullObject pattern, that allows to remove all those cases when using it as it always return valid value for augeas calls.